pkgs/development/compilers/gcc/common/configure-flags.nix

```diff
@@ -114,6 +114,14 @@ let
     ++ lib.optional (!(crossMingw && crossStageStatic))
       "--with-native-system-header-dir=${lib.getDev stdenv.cc.libc}/include"
 
+    # cross-built gcc needs to be tricked as if it is built with headers
+    # available in sysroot. For that we need to pass some --with-sysroot= value:
+    #   https://github.com/NixOS/nixpkgs/pull/181802#issuecomment-1186822355
+    # Without this flag --with-native-system-header-dir= has no effect
+    # and target libc features are not detected like stack protector support
+    ++ lib.optional (buildPlatform != hostPlatform && hostPlatform == targetPlatform)
```

Comment:
Why is the "trick" needed for a cross-built self-targeted compiler, but not needed for canadian cross?

Reply:
I'll answer my own question here: because gcc's autoconfery is kind of a pain.  There are totally separate codepaths for `host==target` and `host!=target`.  Cross-built-native-compilers are not simply a special case of canadian cross compilers.

Reply:
Sounds good!